experiments/yarn/src/main/scala/io/gearpump/experiments/yarn/client/Client.scala

```diff
@@ -17,323 +17,171 @@
+    "name" ->CLIOption[String]("<Application name showed in YARN>", required = false, defaultValue = Some("Gearpump")),
+    "verbose" -> CLIOption("<print verbose log on console>", required = false, defaultValue = Some(false)),
+    "storeConfig" -> CLIOption("<local path where the master config should be stored>", required = false, defaultValue = None)
   )
 
-  val parseResult: ParseResult = parse(args)
-  val config = ConfigFactory.parseFile(new File(parseResult.getString("config")))
+  override def main(akkaConf: Config, args: Array[String]): Unit = {
+    val parsed = parse(args)
 
-  val verbose = parseResult.getBoolean("verbose")
-  if (verbose) {
-    LogUtil.verboseLogToConsole
-  }
+    val name = parsed.getString("name")
+    if (parsed.getBoolean("verbose")) {
+      LogUtil.verboseLogToConsole
+    }
 
-  def apply(appConfig: AppConfig  = new AppConfig(parseResult, config), conf: YarnConfiguration  = new YarnConfiguration,
-            client: YarnClient  = YarnClient.createYarnClient) = {
-    new Client(appConfig, conf, client, ContainerLaunchContext(conf, appConfig), FileSystem.get(conf)).deploy()
-  }
+    val yarnConfig = new YarnConfig()
 
-  def apply(appConfig: AppConfig, conf: YarnConfiguration, client: YarnClient,
-            containerLaunchContext: (String) => ContainerLaunchContext, fileSystem: FileSystem) = {
-    new Client(appConfig, conf, client, containerLaunchContext, fileSystem).deploy()
+    val packagePath = if (parsed.exists("launch")) {
+      parsed.getString("launch")
+    } else {
+      akkaConf.getString(Constants.PACKAGE_PATH)
+    }
+
+    val yarnClient = new YarnClient(yarnConfig, akkaConf)
+
+    val fs = FileSystem.get(yarnConfig.conf)
+    val client = new Client(name, akkaConf, yarnConfig, yarnClient, fs, packagePath)
+
+    Try {
+      val report = client.submit()
+
+      if (parsed.exists("storeConfig")) {
+        val storeFile = new File(parsed.getString("storeConfig"))
+        val configUrl = s"${report.getOriginalTrackingUrl}/api/v1.0/master/config"
+        storeConfig(configUrl, storeFile)
+      }
+
+    }.recover {
+      case ex: FileNotFoundException =>
+        Console.err.println (s"${
+          ex.getMessage
+        }\n" +
+          s"try to check if your gearpump version is right " +
+          s"or HDFS was correctly configured.")
+      case ex => help;
+        throw ex
+    }
   }
 
-  Try(apply()).recover{
-    case ex:FileNotFoundException =>
-      Console.err.println(s"${ex.getMessage}\n" +
-        s"try to check if your gearpump version is right " +
-        s"or HDFS was correctly configured.")
-    case ex => help; throw ex
+  private def storeConfig(url: String, targetFile: File): Unit = {
+    val client = new HttpClient()
+    val get = new GetMethod(url)
+    val status = client.executeMethod(get)
```

Comment:
Currently it looks like the YARN applicationmaster is launched and runs but no /tmp/application.conf is stored when using the -storeConfig option. I believe this is because the services process is not running yet and returns a bad status code. Below I show the appid is returned by the yarnclient and the subdirectory is created with gear.conf, log4j.properties and yarn-site.xml.
```
$ output/target/pack/bin/yarnclient -launch /user/gearpump/gearpump-pack-2.11.5-0.7.1-SNAPSHOT.zip -storeConfig /tmp/application.conf
......================================================
===Application Id: application_1449749298938_5733
Help: 
io.gearpump.experiments.yarn.client.Client -launch <Please specify the gearpump.zip package path on HDFS. If not specified, we will use default value /user/gearpump/gearpump.zip> -name <Application name showed in YARN> -verbose <print verbose log on console> -storeConfig <local path where the master config should be stored> 
-launch  (required:false, default:)
-name  (required:false, default:Gearpump)
-verbose  (required:false, default:false)
-storeConfig  (required:false, default:)
```
```
$ yarn application -list
15/12/13 14:46:56 INFO client.RMProxy: Connecting to ResourceManager at cdh-master-0.node.demogotapaas.consul/10.10.10.146:8032
Total number of applications (application-types: [] and states: [SUBMITTED, ACCEPTED, RUNNING]):2
                Application-Id	    Application-Name	    Application-Type	      User	     Queue	             State	       Final-State	       Progress	                       Tracking-URL
application_1449749298938_5733	            Gearpump	                YARN	  ec2-user	root.ec2-user	           RUNNING	         UNDEFINED	            50%	http://cdh-worker-4.node.demogotapaas.consul:56663
```
```
$ hdfs dfs -ls /user/ec2-user
Found 2 items
drwx------   - ec2-user supergroup          0 2015-12-13 09:28 /user/ec2-user/.Trash
drwx------   - ec2-user supergroup          0 2015-12-13 14:46 /user/ec2-user/.gearpump_app5733
[ec2-user@cdh-master-0 gearpump]$ hdfs dfs -ls /user/ec2-user/.gearpump_app5733
Found 1 items
drwx------   - ec2-user supergroup          0 2015-12-13 14:46 /user/ec2-user/.gearpump_app5733/conf
[ec2-user@cdh-master-0 gearpump]$ hdfs dfs -ls /user/ec2-user/.gearpump_app5733/conf
Found 3 items
-rw-r--r--   3 ec2-user supergroup     166935 2015-12-13 14:46 /user/ec2-user/.gearpump_app5733/conf/gear.conf
-rw-r--r--   3 ec2-user supergroup       2062 2015-12-13 14:46 /user/ec2-user/.gearpump_app5733/conf/log4j.properties
-rw-r--r--   3 ec2-user supergroup      94058 2015-12-13 14:46 /user/ec2-user/.gearpump_app5733/conf/yarn-site.xml
```
```
$ ls /tmp/application.conf
ls: cannot access /tmp/application.conf: No such file or directory
```